README.txt
= rake-remote_task

home :: https://github.com/seattlerb/rake-remote_task
rdoc :: http://docs.seattlerb.org/rake-remote_task

== DESCRIPTION:

Vlad the Deployer's sexy brainchild is rake-remote_task, extending
Rake with remote task goodness.

== FEATURES/PROBLEMS:

* Run remote commands on one or more servers.
* Mix and match local and remote tasks.
* Uses ssh with your ssh settings already in place.
* Uses rsync for efficient transfers.

== EXAMPLE:

  require 'rake/remote_task'

  set :domain, 'abc.example.com'

  remote_task :foo do
    run "ls"
  end

== EXAMPLE SSH CUSTOMIZATION:

To set the ssh command location:

  set :ssh_cmd, '/usr/local/bin/ssh'

To set ssh flags for the login and port:

  set :ssh_flags, %w[-l joe -p 2000]

== SYNOPSIS:

  remote_task :setup_app, :roles => :app do
    # ...
    run "umask #{umask} && mkdir -p #{dirs.join(' ')}"
  end

== REQUIREMENTS:

* rake
* open4 gem (on MRI ruby; not needed for jruby)

== INSTALL:

* sudo gem install rake-remote_task

== LICENSE:

(The MIT License)

Copyright (c) Ryan Davis, RubyHitSquad

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 2.4.4 / 2020-12-22

* 1 bug fix:

  * Bumped rake version. (nearapogee)

=== 2.4.3 / 2018-03-17

* 1 minor enhancement:

  * Extend to allow rake v 12. (piclane)

=== 2.4.2 / 2017-02-27

* 1 bug fix:

  * Fixed 2.4+ Fixnum deprecation warning. (salemine)

=== 2.4.1 / 2016-10-09

* 1 minor enhancement:

  * Extend to include rake 11 (purcell)

=== 2.4.0 / 2016-08-01

* 1 minor enhancement:

  * Add optional yield to run. (bhenderson)

=== 2.3.1 / 2016-05-16

* 1 bug fix:

  * default_env and env were sharing collections. Now duped properly. (bhenderson)

=== 2.3.0 / 2015-06-25

* 1 minor enhancement:

  * Create private methods from Rake::RemoteTask#set. (drbrain)

=== 2.2.2 / 2014-06-17

* 1 minor enhancement:

  * Updated Rake::TestCase for minitest 5.

* 2 bug fixes:

  * Fixed unused variable warnings
  * Removed dead rubyforge setting in Rakefile

=== 2.2.1 / 2013-03-18

* 1 minor enhancement:

  * Pushed revision (head) variable down to vlad. (jari-dk)

=== 2.2.0 / 2013-02-07

* 2 minor enhancements:

  * Fix the command to list the releases. The goal is to reduce the impact of an ssh error message (from off by word count to off by line count)
  * Remove ssh error messages and ancillary dirs from release list

* 1 bug fix:

  * Fixed error with untimetstampted virgin deploy. (goosetav)

=== 2.1.0 / 2012-11-23

* 4 minor enhancements:

  * Allow a set of preliminary commands that will always be run
  * Allow the manipultation of array variable in which one can append values
  * Trust remote umask by default
  * Updated rake dep to be >= 0.8 and < 11.0 for the new wack version scheme. (MichaelXavier)

* 3 bug fixes:

  * Add documentation for +command_prefix+ in +run+
  * Declare append as an operation of Rake:RemoteTask. Fix 7863.
  * Fix misc coding style

=== 2.0.6 / 2012-01-04

* 1 minor enhancement:

  * Add simple SSH example to the readme. (SixArm)

* 1 bug fix:

  * Fix concurrent modification in test with real threads (dekellum)

=== 2.0.5 / 2011-10-14

* 1 minor enhancement:

  * Updated open4 dependency to 1.x family

=== 2.0.4 / 2011-06-22

* 2 bug fixes:

  * Switched rake dependency to ~> 0.8 so we don't collide with other libs
  * Included Rake::DSL inside of Rake::TestCase.

=== 2.0.3 / 2011-02-18

* 2 minor enhancements:

  * Make failure code available via CommandFailedError. (ameuret)
  * Support owner/group in vlad. (RichGuk)

=== 2.0.2 / 2010-09-02

* 1 bug fix:

  * 1.9 fixes to quell test warnings and the like

=== 2.0.1 / 2010-07-23

* 1 bug fix:

  * Fixed syntax error caused by being anal about 80 col boundary (bleything).

=== 2.0.0 / 2010-07-15

* 1 major enhancement

  * Split out from Vlad.




lib/rake/test_case.rb
require 'minitest/autorun'
require 'rake'
require 'rake/remote_task'
require 'stringio'

class StringIO
  def readpartial(size) read end # suck!
end

module Process
  def self.expected status
    @@expected ||= []
    @@expected << status
  end

  class << self
    alias :waitpid2_old :waitpid2

    def waitpid2(pid)
      [ @@expected.shift ]
    end
  end
end

class Rake::RemoteTask
  attr_accessor :commands, :action, :input, :output, :error

  Status = Struct.new :exitstatus

  class Status
    def success?() exitstatus == 0 end
  end

  def system *command
    @commands << command
    self.action ? self.action[command.join(' ')] : true
  end

  def popen4 *command
    @commands << command

    @input = StringIO.new
    out = StringIO.new @output.shift.to_s
    err = StringIO.new @error.shift.to_s

    raise if block_given?

    status = self.action ? self.action[command.join(' ')] : 0
    Process.expected Status.new(status)

    return 42, @input, out, err
  end

  def select reads, writes, errs, timeout
    [reads, writes, errs]
  end
end

class Rake::TestCase < Minitest::Test
  include Rake::DSL if defined? Rake::DSL

  def setup
    @rake = Rake::RemoteTask
    @rake.reset
    Rake.application.clear
    @task_count = Rake.application.tasks.size
    @rake.set :domain, "example.com"
    @lock = Mutex.new
  end

  def util_set_hosts
    @rake.host "app.example.com", :app
    @rake.host "db.example.com", :db
  end

  def util_setup_task(options = {})
    @task = @rake.remote_task :test_task, options
    @task.commands = []
    @task.output   = []
    @task.error    = []
    @task.action   = nil
    @task
  end
end


test/test_rake_remote_task.rb
require 'rake/test_case'

class TestRakeRemoteTask < Rake::TestCase
  def teardown
    Object.send :remove_method, :old_domain if
      Object.private_instance_methods.include? :old_domain
  end

  def test_enhance
    util_set_hosts
    body = Proc.new { 5 }
    task = @rake.remote_task(:some_task => :foo, &body)
    action = Rake::RemoteTask::Action.new(task, body)
    assert_equal [action], task.remote_actions
    assert_equal task, action.task
    assert_equal ["foo"], task.prerequisites
  end

  def test_enhance_with_no_task_body
    util_set_hosts
    util_setup_task
    assert_equal [], @task.remote_actions
    assert_equal [], @task.prerequisites
  end

  def test_execute
    util_set_hosts
    set :some_variable, 1
    set :can_set_nil, nil
    set :lies_are, false
    x = 5
    task = @rake.remote_task(:some_task) do
      @lock.synchronize do
        x += some_variable
      end
    end
    task.execute nil
    assert_equal 1, task.send(:some_variable)
    assert_equal 7, x
    assert task.send(:can_set_nil).nil?
    assert_equal false, task.send(:lies_are)
  end

  def test_set_false
    set :can_set_nil, nil
    set :lies_are, false

    assert_nil task.send(:can_set_nil)

    assert_equal false, task.send(:lies_are)
    assert_equal false, Rake::RemoteTask.fetch(:lies_are)
  end


  def test_fetch_false
    assert_equal false, Rake::RemoteTask.fetch(:unknown, false)
  end

  def test_execute_exposes_target_host
    host "app.example.com", :app
    task = remote_task(:target_task) { set(:test_target_host, target_host) }
    task.execute nil
    assert_equal "app.example.com", Rake::RemoteTask.fetch(:test_target_host)
  end

  def test_execute_with_no_hosts
    @rake.host "app.example.com", :app
    t = @rake.remote_task(:flunk, :roles => :db) { flunk "should not have run" }
    e = assert_raises(Rake::ConfigurationError) { t.execute nil }
    assert_equal "No target hosts specified on task flunk for roles [:db]",
                 e.message
  end

  def test_execute_with_no_roles
    t = @rake.remote_task(:flunk, :roles => :junk) { flunk "should not have run" }
    e = assert_raises(Rake::ConfigurationError) { t.execute nil }
    assert_equal "No target hosts specified on task flunk for roles [:junk]",
                 e.message
  end

  def test_execute_with_roles
    util_set_hosts
    set :some_variable, 1
    x = 5
    task = @rake.remote_task(:some_task, :roles => :db) { x += some_variable }
    task.execute nil
    assert_equal 1, task.send(:some_variable)
    assert_equal 6, x
  end

  def test_rsync
    util_setup_task
    @task.target_host = "app.example.com"

    assert_silent do
      @task.rsync 'localfile', 'host:remotefile'
    end

    commands = @task.commands

    assert_equal 1, commands.size, 'not enough commands'
    assert_equal(%w[rsync -azP --delete localfile host:remotefile],
                 commands.first)
  end

  def test_rsync_fail
    util_setup_task
    @task.target_host = "app.example.com"
    @task.action = proc { false }

    e = nil
    assert_silent do
      e = assert_raises Rake::CommandFailedError do
        @task.rsync 'local', 'host:remote'
      end
    end
    exp = "execution failed: rsync -azP --delete local host:remote"
    assert_equal exp, e.message
  end

  def test_rsync_deprecation
    util_setup_task
    @task.target_host = "app.example.com"

    out, err = capture_io do
      @task.rsync 'localfile', 'remotefile'
    end

    commands = @task.commands

    assert_equal 1, commands.size, 'not enough commands'
    assert_equal(%w[rsync -azP --delete localfile app.example.com:remotefile],
                 commands.first)

    assert_equal("rsync deprecation: pass target_host:remote_path explicitly\n",
                 err)
    assert_empty out
    # flunk "not yet"
  end

  def test_get
    util_setup_task
    @task.target_host = "app.example.com"

    assert_silent do
      @task.get 'tmp', "remote1", "remote2"
    end

    commands = @task.commands

    expected = %w[rsync -azP --delete app.example.com:remote1 app.example.com:remote2 tmp]

    assert_equal 1, commands.size
    assert_equal expected, commands.first
  end

  def test_put
    util_setup_task
    @task.target_host = "app.example.com"

    assert_silent do
      @task.put 'dest' do
        "whatever"
      end
    end

    commands = @task.commands

    expected  = %w[rsync -azP --delete HAPPY app.example.com:dest]
    commands.first[3] = 'HAPPY'

    assert_equal 1, commands.size
    assert_equal expected, commands.first
  end

  def test_run
    util_setup_task
    @task.output << "file1\nfile2\n"
    @task.target_host = "app.example.com"
    result = nil

    out, err = capture_io do
      result = @task.run("ls")
    end

    commands = @task.commands

    assert_equal 1, commands.size, 'not enough commands'
    assert_equal ["ssh", "app.example.com", "ls"],
                 commands.first, 'app'
    assert_equal "file1\nfile2\n", result

    assert_equal "file1\nfile2\n", out
    assert_equal '', err
  end

  def test_run_dir
    util_setup_task
    @task.target_host = "app.example.com:/www/dir1"

    @task.run("ls")

    commands = @task.commands

    assert_equal 1, commands.size, 'not enough commands'
    assert_equal [["ssh", "app.example.com", "cd /www/dir1 && ls"]], commands
  end

  def test_run_failing_command
    util_set_hosts
    util_setup_task
    @task.input = StringIO.new "file1\nfile2\n"
    @task.target_host =  'app.example.com'
    @task.action = proc { 1 }

    e = assert_raises(Rake::CommandFailedError) { @task.run("ls") }
    assert_equal "execution failed with status 1: ssh app.example.com ls", e.message

    assert_equal 1, @task.commands.size
  end

  def test_run_sudo
    util_setup_task
    @task.output << "file1\nfile2\n"
    @task.error << 'Password:'
    @task.target_host = "app.example.com"
    def @task.sudo_password() "my password" end # gets defined by set
    result = nil

    out, err = capture_io do
      result = @task.run("sudo ls")
    end

    commands = @task.commands

    assert_equal 1, commands.size, 'not enough commands'
    assert_equal ['ssh', 'app.example.com', 'sudo ls'],
                 commands.first

    assert_equal "my password\n", @task.input.string

    # WARN: Technically incorrect, the password line should be
    # first... this is an artifact of changes to the IO code in run
    # and the fact that we have a very simplistic (non-blocking)
    # testing model.
    assert_equal "file1\nfile2\nPassword:\n", result

    assert_equal "file1\nfile2\n", out
    assert_equal "Password:\n", err
  end

  def test_run_with_block
    util_setup_task
    @task.output << "file1\nfile2\n"
    @task.error << 'Enter Password:'
    @task.target_host = "app.example.com"
    result = nil

    out, err = capture_io do
      result = @task.run("./command/with/input") do |input, stream, data|
        if stream == :out and data == "file1\nfile2\n"
          input << "myinput\n"
        elsif stream == :err and data == "Enter Password:"
          input << "mysecret\n"
        end
      end
    end

    commands = @task.commands

    assert_equal 1, commands.size, 'not enough commands'
    assert_equal ['ssh', 'app.example.com', './command/with/input'],
                 commands.first

    assert_equal "myinput\nmysecret\n", @task.input.string

    # WARN: Technically incorrect, the password line should be
    # first... this is an artifact of changes to the IO code in run
    # and the fact that we have a very simplistic (non-blocking)
    # testing model.
    assert_equal "file1\nfile2\nEnter Password:", result

    assert_equal "file1\nfile2\n", out
    assert_equal "Enter Password:", err
  end

  def test_sudo
    util_setup_task
    @task.target_host = "app.example.com"
    @task.sudo "ls"

    commands = @task.commands

    assert_equal 1, commands.size, 'wrong number of commands'
    assert_equal ["ssh", "app.example.com", "sudo -p Password: ls"],
                 commands.first, 'app'
  end

  def test_append
    append :some_array, 1
    assert_equal [1], some_array
  end
end


lib/rake/remote_task/open3.rb
require 'open3'

##
# Implementation of +run+ using stdlib Open3. Currently this is only
# viable on JRuby where the exit status may be obtained. By comparison
# open4 requires a fork which isn't available on JRuby.
module Rake::RemoteOpen3
  include Open3

  ##
  # Use ssh to execute +command+ on target_host. If +command+ uses
  # sudo, the sudo password will be prompted for then saved for
  # subsequent sudo commands.
  def run command
    command = "cd #{target_dir} && #{command}" if target_dir
    cmd     = [ssh_cmd, ssh_flags, target_host, command].flatten
    result  = []

    trace = [ssh_cmd, ssh_flags, target_host, "'#{command}'"].flatten.join(' ')
    warn trace if $TRACE

    popen3(*cmd) do |inn,out,err|

      inn.sync   = true
      streams    = [out, err]
      out_stream = {
        out => $stdout,
        err => $stderr,
      }

      until streams.empty? do
        # don't busy loop
        selected, = select(streams, nil, nil, 0.1)

        next if selected.nil? or selected.empty?

        selected.each do |stream|
          if stream.eof? then
            streams.delete stream
            next
          end

          data = stream.readpartial(1024)
          out_stream[stream].write data

          if stream == err and data =~ sudo_prompt then
            inn.puts sudo_password
            data << "\n"
            $stderr.write "\n"
          end

          result << data
        end
      end

      # Make sure an open input stream doesn't keep the process
      # running. The others are closed on exit from block by popen3
      # itself.
      inn.close rescue nil
    end

    # MRI popen3 uses double fork and thus has no way to set `$?`, see:
    #
    #   http://bugs.ruby-lang.org/issues/1287
    #
    # However the `$?` is available after the popen3 {} (block form
    # only) in the non-forking jruby implementation. Good thing since
    # waitpid* isn't reliable here.
    #
    # test_status will be non-null in test
    status = test_status || $?

    unless status.exitstatus == 0 then
      raise(Rake::CommandFailedError.new(status),
            "execution failed with status #{status.exitstatus}: #{cmd.join ' '}")
    end

    result.join
  end

  private

  ##
  # Test hook for injecting status (can't set $?)
  def test_status
    nil
  end

end


lib/rake/remote_task.rb
require 'rubygems'
require 'open4'
require 'rake'

$TESTING ||= false
$TRACE = Rake.application.options.trace
$-w = true if $TRACE # asshat, don't mess with my warn.

[
 ["Thread.current[:task]", :get, :put, :rsync, :run, :sudo, :target_host],
 ["Rake::RemoteTask",      :host, :remote_task, :role, :set, :append]
].each do |methods|
  receiver = methods.shift
  methods.each do |method|
    eval "def #{method} *args, &block; #{receiver}.#{method}(*args, &block);end"
  end
end

module Rake
  ##
  # Base error class for all Vlad errors.
  class Error < RuntimeError; end

  ##
  # Raised when you have incorrectly configured Vlad.
  class ConfigurationError < Error; end

  ##
  # Raised when a remote command fails.
  class CommandFailedError < Error
    attr_reader :status
    def initialize( status )
      @status = status
    end
  end

  ##
  # Raised when an environment variable hasn't been set.
  class FetchError < Error; end
end

##
# Rake::RemoteTask is a subclass of Rake::Task that adds
# remote_actions that execute in parallel on multiple hosts via ssh.

class Rake::RemoteTask < Rake::Task

  VERSION = "2.4.4"

  @@current_roles = []

  include Open4

  ##
  # Options for execution of this task.

  attr_accessor :options

  ##
  # The host this task is running on during execution.

  attr_reader :target_host

  ##
  # The directory on the host this task is running in during execution.

  attr_reader :target_dir

  ##
  # An Array of Actions this host will perform during execution. Use
  # enhance to add new actions to a task.

  attr_reader :remote_actions

  def self.current_roles
    @@current_roles
  end

  ##
  # Create a new task named +task_name+ attached to Rake::Application +app+.

  def initialize(task_name, app)
    super

    @remote_actions = []
    @happy = false # used for deprecation warnings on get/put/rsync
  end

  ##
  # Add a local action to this task. This calls Rake::Task#enhance.

  alias_method :original_enhance, :enhance

  ##
  # Add remote action +block+ to this task with dependencies +deps+. See
  # Rake::Task#enhance.

  def enhance(deps=nil, &block)
    original_enhance(deps) # can't use super because block passed regardless.
    @remote_actions << Action.new(self, block) if block_given?
    self
  end

  ##
  # Execute this action. Local actions will be performed first, then remote
  # actions will be performed in parallel on each host configured for this
  # RemoteTask.

  def execute(args = nil)
    raise(Rake::ConfigurationError,
          "No target hosts specified on task #{self.name} for roles #{options[:roles].inspect}") unless
      defined_target_hosts?

    super args

    @remote_actions.each { |act| act.execute(target_hosts, self, args) }
  end

  ##
  # Pull +files+ from the remote +host+ using rsync to +local_dir+.
  # TODO: what if role has multiple hosts & the files overlap? subdirs?

  def get local_dir, *files
    @happy = true
    host = target_host
    rsync files.map { |f| "#{host}:#{f}" }, local_dir
    @happy = false
  end

  ##
  # Copy a (usually generated) file to +remote_path+. Contents of block
  # are copied to +remote_path+ and you may specify an optional
  # base_name for the tempfile (aids in debugging).

  def put remote_path, base_name = File.basename(remote_path)
    require 'tempfile'
    Tempfile.open base_name do |fp|
      fp.puts yield
      fp.flush
      @happy = true
      rsync fp.path, "#{target_host}:#{remote_path}"
      @happy = false
    end
  end

  ##
  # Execute rsync with +args+. Tacks on pre-specified +rsync_cmd+ and
  # +rsync_flags+.
  #
  # Favor #get and #put for most tasks. Old-style direct use where the
  # target_host was implicit is now deprecated.

  def rsync *args
    unless @happy || args[-1] =~ /:/ then
      warn "rsync deprecation: pass target_host:remote_path explicitly"
      args[-1] = "#{target_host}:#{args[-1]}"
    end

    cmd    = [rsync_cmd, rsync_flags, args].flatten.compact
    cmdstr = cmd.join ' '

    warn cmdstr if $TRACE

    success = system(*cmd)

    raise Rake::CommandFailedError.new($?), "execution failed: #{cmdstr}" unless success
  end

  ##
  # Use ssh to execute +command+ on target_host. If +command+ uses sudo, the
  # sudo password will be prompted for then saved for subsequent sudo commands.
  #
  # If +command_prefix+ has been filled up with one or several commands, they
  # will be run on target_host before +command+.
  #
  # Yields input channel, :out or :err, and data.

  def run command
    commands = []

    commands << "cd #{target_dir}" if target_dir
    commands += command_prefix.flatten if command_prefix
    commands << "#{command}"

    command = commands.join(" && ")

    cmd     = [ssh_cmd, ssh_flags, target_host, command].flatten
    result  = []

    trace = [ssh_cmd, ssh_flags, target_host, "'#{command}'"].flatten.join(' ')
    warn trace if $TRACE

    pid, inn, out, err = popen4(*cmd)

    inn.sync   = true
    streams    = [out, err]
    out_stream = {
      out => $stdout,
      err => $stderr,
    }

    # Handle process termination ourselves
    status = nil
    Thread.start do
      status = Process.waitpid2(pid).last
    end

    until streams.empty? do
      # don't busy loop
      selected, = select streams, nil, nil, 0.1

      next if selected.nil? or selected.empty?

      selected.each do |stream|
        if stream.eof? then
          streams.delete stream if status # we've quit, so no more writing
          next
        end

        data = stream.readpartial(1024)
        out_stream[stream].write data

        if stream == err and data =~ sudo_prompt then
          inn.puts sudo_password
          data << "\n"
          $stderr.write "\n"
        end

        yield inn, stream == out ? :out : :err, data if block_given?

        result << data
      end
    end

    unless status.success? then
      raise(Rake::CommandFailedError.new(status),
            "execution failed with status #{status.exitstatus}: #{cmd.join ' '}")
    end

    result.join
  ensure
    inn.close rescue nil
    out.close rescue nil
    err.close rescue nil
  end

  ##
  # Returns an Array with every host configured.

  def self.all_hosts
    hosts_for(roles.keys)
  end

  ##
  # The default environment values. Used for resetting (mostly for
  # tests).

  def self.default_env
    @@default_env
  end

  def self.per_thread
    @@per_thread
  end

  ##
  # The vlad environment.

  def self.env
    @@env
  end

  def self.is_array
    @@is_array
  end

  ##
  # Fetches environment variable +name+ from the environment using
  # default +default+.

  def self.fetch name, default = nil
    name = name.to_s if Symbol === name
    if @@env.has_key? name then
      protect_env(name) do
        v = @@env[name]
        if @@is_array[name] then
          v = v.map do |item|
            Proc === item ? item.call : item
          end
          unless per_thread[name] then
            @@env[name] = v
            @@is_array[name] = false
          end
        elsif Proc === v then
          v = v.call
          @@env[name] = v unless per_thread[name]
        end
        v
      end
    elsif default || default == false then
      @@env[name] = default
    else
      raise Rake::FetchError
    end
  end

  ##
  # Add host +host_name+ that belongs to +roles+. Extra arguments may
  # be specified for the host as a hash as the last argument.
  #
  # host is the inversion of role:
  #
  #   host 'db1.example.com', :db, :master_db
  #
  # Is equivalent to:
  #
  #   role :db, 'db1.example.com'
  #   role :master_db, 'db1.example.com'

  def self.host host_name, *roles
    opts = Hash === roles.last ? roles.pop : {}

    roles.each do |role_name|
      role role_name, host_name, opts.dup
    end
  end

  ##
  # Returns an Array of all hosts in +roles+.

  def self.hosts_for *roles
    roles.flatten.map { |r|
      self.roles[r].keys
    }.flatten.uniq.sort
  end

  def self.mandatory name, desc # :nodoc:
    self.set(name) do
      raise(Rake::ConfigurationError,
            "Please specify the #{desc} via the #{name.inspect} variable")
    end
  end

  ##
  # Ensures exclusive access to +name+.

  def self.protect_env name # :nodoc:
    @@env_locks[name].synchronize do
      yield
    end
  end

  ##
  # Adds a remote task named +name+ with options +options+ that will
  # execute +block+.

  def self.remote_task name, *args, &block
    options = (Hash === args.last) ? args.pop : {}
    t = Rake::RemoteTask.define_task(name, *args, &block)
    options[:roles] = Array options[:roles]
    options[:roles] |= @@current_roles
    t.options = options
    t
  end

  ##
  # Ensures +name+ does not conflict with an existing method.

  def self.reserved_name? name # :nodoc:
    !@@env.has_key?(name.to_s) && self.respond_to?(name)
  end

  ##
  # Resets vlad, restoring all roles, tasks and environment variables
  # to the defaults.

  def self.reset
    @@def_role_hash = {}                # official default role value
    @@env           = {}
    @@is_array      = {}
    @@tasks         = {}
    @@roles         = Hash.new { |h,k| h[k] = @@def_role_hash }
    @@env_locks     = Hash.new { |h,k| h[k] = Mutex.new }

    @@default_env.each do |k,v|
      @@env[k] = safe_dup(v)
    end
  end

  def self.safe_dup v # :nodoc:
    case v
    when Symbol, Integer, nil, true, false, 42 then # ummmm... yeah. bite me.
      v
    else
      v.dup
    end
  end

  ##
  # Adds role +role_name+ with +host+ and +args+ for that host.
  # TODO: merge:
  # Declare a role and assign a remote host to it. Equivalent to the
  # <tt>host</tt> method; provided for capistrano compatibility.

  def self.role role_name, host = nil, args = {}
    if block_given? then
      raise ArgumentError, 'host not allowed with block' unless host.nil?

      begin
        current_roles << role_name
        yield
      ensure
        current_roles.delete role_name
      end
    else
      raise ArgumentError, 'host required' if host.nil?

      [*host].each do |hst|
        raise ArgumentError, "invalid host: #{hst}" if hst.nil? or hst.empty?
      end
      @@roles[role_name] = {} if @@def_role_hash.eql? @@roles[role_name]
      @@roles[role_name][host] = args
    end
  end

  ##
  # The configured roles.

  def self.roles
    host domain, :app, :web, :db if @@roles.empty?

    @@roles
  end

  ##
  # Set environment variable +name+ to +value+ or +default_block+.
  #
  # If +default_block+ is defined, the block will be executed the
  # first time the variable is fetched, and the value will be used for
  # every subsequent fetch.

  def self.set name, value = nil, &default_block
    raise ArgumentError, "cannot provide both a value and a block" if
      value and default_block unless
      value == :per_thread
    raise ArgumentError, "cannot set reserved name: '#{name}'" if
      Rake::RemoteTask.reserved_name?(name) unless $TESTING

    name = name.to_s

    Rake::RemoteTask.per_thread[name] = true if
      default_block && value == :per_thread

    Rake::RemoteTask.default_env[name] = default_block || safe_dup(value)
    Rake::RemoteTask.env[name]         = default_block || safe_dup(value)

    if Object.private_instance_methods.include? name.to_sym then
      Object.send :alias_method, :"old_#{name}", name
    end

    Object.send :define_method, name do
      Rake::RemoteTask.fetch name
    end

    Object.send :private, name
  end

  ##
  # Append +value+ or +default_block+ to  environment variable +name+
  #
  # To initialize an empty array, just do append +name+
  #
  # If +default_block+ is defined, the block will be executed the
  # first time the variable is fetched, and the value will be used for
  # every subsequent fetch.

  def self.append name, value = nil, &default_block
    raise ArgumentError, "cannot provide both a value and a block" if
      value and default_block unless
      value == :per_thread
    raise ArgumentError, "cannot set reserved name: '#{name}'" if
      Rake::RemoteTask.reserved_name?(name) unless $TESTING

    name = name.to_s

    set(name, []) unless @@is_array[name]
    Rake::RemoteTask.is_array[name] = true
    Rake::RemoteTask.per_thread[name] ||= default_block && value == :per_thread

    v = default_block || value
    if v then
      Rake::RemoteTask.default_env[name] << v
      Rake::RemoteTask.env[name] << v
    end
  end

  ##
  # Sets all the default values. Should only be called once. Use reset
  # if you need to restore values.

  def self.set_defaults
    @@default_env ||= {}
    @@per_thread  ||= {}
    self.reset

    mandatory :repository, "repository path"
    mandatory :deploy_to,  "deploy path"
    mandatory :domain,     "server domain"

    simple_set(:deploy_timestamped, true,
               :deploy_via,         :export,
               :keep_releases,      5,
               :rake_cmd,           "rake",
               :rsync_cmd,          "rsync",
               :rsync_flags,        ['-azP', '--delete'],
               :ssh_cmd,            "ssh",
               :ssh_flags,          [],
               :sudo_cmd,           "sudo",
               :sudo_flags,         ['-p Password:'],
               :sudo_prompt,        /^Password:/,
               :umask,              nil,
               :mkdirs,             [],
               :shared_paths,       {},
               :perm_owner,         nil,
               :perm_group,         nil)

    append :command_prefix, []

    set(:current_release)    { (releases.any?) ? File.join(releases_path, releases[-1]) : nil }
    set(:latest_release)     {
      deploy_timestamped ? release_path : current_release
    }
    set(:previous_release)   { File.join(releases_path, releases[-2]) }
    set(:release_name)       { Time.now.utc.strftime("%Y%m%d%H%M%S") }
    set(:release_path)       { File.join(releases_path, release_name) }
    set(:releases)           { task.run("ls -1 #{releases_path}").split("\n").select { |l| l =~ /^\d+$/ }.sort }

    set_path :current_path,  "current"
    set_path :releases_path, "releases"
    set_path :scm_path,      "scm"
    set_path :shared_path,   "shared"

    set(:sudo_password) do
      state = `stty -g`

      raise Rake::Error, "stty(1) not found" unless $?.success?

      begin
        system "stty -echo"
        $stdout.print "sudo password: "
        $stdout.flush
        sudo_password = $stdin.gets
        $stdout.puts
      ensure
        system "stty #{state}"
      end
      sudo_password
    end
  end

  def self.set_path(name, subdir) # :nodoc:
    set(name) { File.join(deploy_to, subdir) }
  end

  def self.simple_set(*args) # :nodoc:
    args = Hash[*args]
    args.each do |k, v|
      set k, v
    end
  end

  ##
  # The Rake::RemoteTask executing in this Thread.

  def self.task
    Thread.current[:task]
  end

  ##
  # The configured Rake::RemoteTasks.

  def self.tasks
    @@tasks
  end

  ##
  # Execute +command+ under sudo using run.

  def sudo command
    run [sudo_cmd, sudo_flags, command].flatten.compact.join(" ")
  end

  ##
  # Sets the target host. Allows you to set an optional directory
  # using the format:
  #
  #    host.domain:/dir

  def target_host= host
    if host =~ /^(.+):(.+?)$/
      @target_host = $1
      @target_dir  = $2
    else
      @target_host = host
      @target_dir  = nil
    end
  end

  ##
  # The hosts this task will execute on. The hosts are determined from
  # the role this task belongs to.
  #
  # The target hosts may be overridden by providing a comma-separated
  # list of commands to the HOSTS environment variable:
  #
  #   rake my_task HOSTS=app1.example.com,app2.example.com

  def target_hosts
    if hosts = ENV["HOSTS"] then
      hosts.strip.gsub(/\s+/, '').split(",")
    else
      roles = Array options[:roles]

      if roles.empty? then
        Rake::RemoteTask.all_hosts
      else
        Rake::RemoteTask.hosts_for roles
      end
    end
  end

  ##
  # Similar to target_hosts, but returns true if user defined any hosts, even
  # an empty list.

  def defined_target_hosts?
    return true if ENV["HOSTS"]
    roles = Array options[:roles]
    return true if roles.empty?
    # borrowed from hosts_for:
    roles.flatten.each { |r|
      return true unless @@def_role_hash.eql? Rake::RemoteTask.roles[r]
    }
    return false
  end

  ##
  # Action is used to run a task's remote_actions in parallel on each
  # of its hosts. Actions are created automatically in
  # Rake::RemoteTask#enhance.

  class Action

    ##
    # The task this action is attached to.

    attr_reader :task

    ##
    # The block this action will execute.

    attr_reader :block

    ##
    # An Array of threads, one for each host this action executes on.

    attr_reader :workers

    ##
    # Creates a new Action that will run +block+ for +task+.

    def initialize task, block
      @task  = task
      @block = block
      @workers = ThreadGroup.new
    end

    def == other # :nodoc:
      return false unless Action === other
      block == other.block && task == other.task
    end

    ##
    # Execute this action on +hosts+ in parallel. Returns when block
    # has completed for each host.

    def execute hosts, task, args
      hosts.each do |host|
        t = task.clone
        t.target_host = host
        thread = Thread.new(t) do |task2|
          Thread.current[:task] = task2
          case block.arity
          when 1
            block.call task2
          else
            block.call task2, args
          end
          Thread.current[:task] = nil
        end
        @workers.add thread
      end
      @workers.list.each { |thr| thr.join }
    end
  end
end

Rake::RemoteTask.set_defaults
